src/main/java/com/bootcamp/demo/restapi/BookingsController.java

```diff
@@ -0,0 +1,178 @@
+    }
+
+    @PostMapping("/bookScooter")
+    @ResponseBody
+    public void bookScooter(final Double lat, final Double longitude) throws ExecutionException, InterruptedException {
+        bookAScooter(lat, longitude);
+
+    }
+
+    @DeleteMapping("/delete")
+    public ResponseEntity<Object> deleteBooking(@RequestParam final String id) {
+        try {
+            LOGGER.info("DELETE BOOKING - API endpoint invoked");
+            bookingService.deleteBooking(id);
+            return new ResponseEntity<>(HttpStatus.OK);
+        } catch (ExecutionException | InterruptedException | IllegalArgumentException e) {
+            LOGGER.info("Unfortunately,an error happened when trying to delete a booking.");
+            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    @PutMapping("/update")
+    public ResponseEntity<Object> updateBooking(@RequestParam final String id, @RequestParam("end")
+            String endDate, PaymentStatus status) {
+        try {
+            LOGGER.info("UPDATE BOOKING - API endpoint invoked");
+
+            Booking booking = bookingService.getBookingByID(id);
+            Scooter scooter = scooterService.findScooterById(booking.getSerialNumber());
+
+            if (status == PaymentStatus.SUCCESS) {
+                endBookingAndUpdate(booking, scooter, endDate);
+            }
+            return new ResponseEntity<>(HttpStatus.OK);
+        } catch (ExecutionException | InterruptedException | IllegalArgumentException e) {
+            LOGGER.info("Unfortunately,an error happened when trying to update the booking.");
+
+            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    public double endBookingAndUpdate(Booking booking, Scooter scooter, String endDate) throws ExecutionException, InterruptedException {
+        long bookingDuration = 0;
+
+        String startDate = booking.getStartDate();
+        //Date endDate = booking.getEndDate();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
+        LocalDateTime dateTimeStart = LocalDateTime.parse(startDate, formatter);
+        LocalDateTime dateTimeEnd = LocalDateTime.parse(endDate, formatter);
+        if (dateTimeEnd.isAfter(dateTimeStart)) {
+            bookingDuration = dateTimeEnd.getMinute() - dateTimeStart.getMinute();
+        }
+        final double totalCostComputed = booking.getTotalCost(bookingDuration);
+        bookingService.updateBooking(booking.getId(), endDate, booking.getPayment());
+        scooter.setCurrentLocation(new Location(scooter.getCurrentLocation().getLongitude(), scooter.getCurrentLocation().getLongitude()));
+        scooterService.updateScooter(scooter.getSerialNumber(), scooter.getCurrentLocation(), ScooterStatus.AVAILABLE, scooter.getBattery().getLevel());
+
+        return totalCostComputed;
+    }
+
+    public Scooter bookAScooter(final Double lat, final Double longitude) throws ExecutionException, InterruptedException {

```

Comment:
This method is very long and a bit hard to follow. For such cases, it is best to split this is multiple smaller methods which do 1 thing. Look at the suggestion below and check if it is easier to follow, read and understand. Also, find there also some comments (which should not be added after refactoring, but I thought it might be easier to follow as such, than adding separate comments and having to take both those & the refactoring suggesting into account) - for simplicity I have added those with preceded with ////.

```
//// no need to be initialized every time
private static final Random RANDOM = new Random();
....
public Scooter bookAScooter(final Double lat, final Double longitude) throws ExecutionException, InterruptedException {
    List<Scooter> allAvailableScooters = scooterService.getAvailableScooters(new Location(lat, longitude), 10.0);
    return !allAvailableScooters.isEmpty()
            ? bookRandomScooter(allAvailableScooters)
            : null;
}

private Scooter bookRandomScooter(List<Scooter> allAvailableScooters) throws ExecutionException, InterruptedException {
    final Scooter pickedScooter = reserveScooter(allAvailableScooters);
    createBooking(pickedScooter);
    return pickedScooter;
}

private void createBooking(Scooter scooter) throws ExecutionException, InterruptedException {
//// Shouldn't here a new booking be created? Or were you thinking that you would first call createBooking & then this one?
//// In my opinion, it might be more intuitive to create the booking here with attributes sth like
////      serialNumber - the one from the scooter
////      accountId - you provide as input the user's id to this method
////      startDate - would be LocalDateTime.now
////      endDate - null
////      payment - null
    final List<Booking> bookings = bookingService.getAllBookings();
    Booking pickedBooking = bookings.get(getRandom(bookings.size()));
    pickedBooking.setSerialNumber(scooter.getSerialNumber());
//// redundant, should be removed
////            String startDateBooking = pickedBooking.getStartDate();
////            pickedBooking.setStartDate(startDateBooking);
    bookingService.createBooking(pickedBooking);
}

private Scooter reserveScooter(final List<Scooter> allAvailableScooters) throws ExecutionException, InterruptedException {
    final Scooter pickedScooter = allAvailableScooters.get(getRandom(allAvailableScooters.size()));
    pickedScooter.setStatus(ScooterStatus.RESERVED);
//// Wondering if here it would be simpler to send the scooter directly.
//// What is the benefit of sending these params & creating a new scooter? Note that if you would use Scooter as param, 
// you could call `createScooter`, since it does the update as well, as it calls set on firestore
    scooterService.updateScooter(pickedScooter.getSerialNumber(), pickedScooter.getCurrentLocation(), pickedScooter.getStatus(), pickedScooter.getBattery().getLevel());
    return pickedScooter;
}

//// should be renamed to be more intuitive
private int getRandom(final int total) {
    return RANDOM.nextInt(total);
}
```